packages/contracts-sdk/src/implementations/farcaster/FarcasterIdGatewayContract.ts

```diff
@@ -0,0 +1,199 @@
+} from "@snickerdoodlelabs/objects";
+import { TypedDataField, ethers } from "ethers";
+import { injectable } from "inversify";
+import { ResultAsync, okAsync } from "neverthrow";
+
+import { BaseContract } from "@contracts-sdk/implementations/BaseContract.js";
+import { IEthersContractError } from "@contracts-sdk/implementations/BlockchainErrorMapper.js";
+import {
+  ContractOverrides,
+  IFarcasterIdGatewayContract,
+  WrappedTransactionResponse,
+} from "@contracts-sdk/interfaces/index.js";
+import { ContractsAbis } from "@contracts-sdk/interfaces/objects/index.js";
+
+@injectable()
+export class FarcasterIdGatewayContract
+  extends BaseContract<FarcasterIdGatewayContractError>
+  implements IFarcasterIdGatewayContract
+{
+  constructor(
+    protected providerOrSigner: ethers.Provider | ethers.Signer,
+    protected contractAddress: EVMContractAddress,
+  ) {
+    super(
+      providerOrSigner,
+      contractAddress,
+      ContractsAbis.FarcasterIdGatewayAbi.abi,
+    );
+  }
+
+  public price(
+    extraStorage?: bigint,
+  ): ResultAsync<
+    bigint,
+    FarcasterIdGatewayContractError | BlockchainCommonErrors
+  > {
+    // If extraStorage was provided, call price with it, otherwise call without
+    // https://optimistic.etherscan.io/address/0x00000000fc25870c6ed6b6c7e41fb078b7656f69#code#F2#L64
+    return ResultAsync.fromPromise(
+      extraStorage
+        ? (this.contract.price(extraStorage) as Promise<bigint>)
+        : (this.contract.price() as Promise<bigint>),
+      (e) => {
+        return this.generateError(e, "Unable to call price()");
+      },
+    );
+  }
+
+  public nonces(
+    address: EVMAccountAddress,
+  ): ResultAsync<
+    bigint,
+    FarcasterIdGatewayContractError | BlockchainCommonErrors
+  > {
+    return ResultAsync.fromPromise(
+      this.contract.nonces(address) as Promise<bigint>,
+      (e) => {
+        return this.generateError(e, "Unable to call nonces()");
+      },
+    );
+  }
+
+  public register(
+    recoveryAddress: EVMAccountAddress,
+    overrides?: ContractOverrides,
+  ): ResultAsync<
+    WrappedTransactionResponse,
+    BlockchainCommonErrors | FarcasterIdGatewayContractError
+  > {
+    return this.writeToContract("register", [recoveryAddress], overrides);
+  }
+
+  public registerWithExtraStorage(
+    recoveryAddress: EVMAccountAddress,
+    extraStorage: bigint,
+    overrides?: ContractOverrides,
+  ): ResultAsync<
+    WrappedTransactionResponse,
+    BlockchainCommonErrors | FarcasterIdGatewayContractError
+  > {
+    return this.writeToContract(
+      "register",
+      [recoveryAddress, extraStorage],
+      overrides,
+    );
+  }
+
+  // Called by Admin EW to register for on behalf of a user's EW
+  public registerFor(
+    ownerAddress: EVMAccountAddress,
+    recoveryAddress: EVMAccountAddress,
+    deadline: UnixTimestamp,
+    signature: SignedRegisterSignature, // EIP-712 'Register' signature from fidOwner obtained from getRegisterSignature()
```

Comment:
SignedRegisterSignature seems redundant, could this be a FarcasterRegisterSignature? Would that read better?